File: list_endpoints.ts
/**
 * list_endpoints — Scan the workspace and return all API endpoints and routes.
 *
 * Detects:
 *   Java/Spring:        @QueryMapping, @MutationMapping, @GetMapping, @PostMapping,
 *                       @PutMapping, @PatchMapping, @DeleteMapping
 *   TypeScript/RR7:     loader, action exports in routes/ files; resource routes under routes/api/
 *
 * Returns a compact index — name, kind, file, line, path — no file content.
 * Modular: new languages can be added by registering an analyzer in endpoints_core/analyzers/.
 *
 * Examples:
 *   list_endpoints()                          — all languages, all types
 *   list_endpoints(language="java")           — Spring endpoints only
 *   list_endpoints(language="java", type="graphql")   — only @QueryMapping / @MutationMapping
 *   list_endpoints(language="java", type="rest")      — only HTTP mappings
 *   list_endpoints(language="typescript")     — React Router 7 routes only
 *   list_endpoints(language="typescript", type="routes") — loaders and actions only
 */

import { tool } from "@opencode-ai/plugin"
import * as path from "node:path"
import * as fs from "node:fs/promises"

const SCRIPT = path.join(
  path.dirname(new URL(import.meta.url).pathname),
  "list_endpoints.py",
)

function uvDepsForLanguage(language: string): string[] {
  const base: string[] = []
  switch (language) {
    case "java":
      return [...base, "--with", "tree-sitter", "--with", "tree-sitter-java"]
    case "typescript":
      return base  // TS analyzer uses regex only — no tree-sitter needed
    case "auto":
    default:
      return [...base, "--with", "tree-sitter", "--with", "tree-sitter-java"]
  }
}

export default tool({
  description: `Scan the workspace and return all API endpoints and frontend routes — no file content.

Returns a compact index: name, kind, file, line, and HTTP path when available.
Use this to understand the full API surface of a project before reading individual files.

Supports:
  - Java / Spring Boot: @QueryMapping, @MutationMapping, @GetMapping, @PostMapping,
                        @PutMapping, @PatchMapping, @DeleteMapping
  - TypeScript / React Router 7: loader, action exports; resource routes under routes/api/

Modular — new languages can be added without changing this tool.

Examples:
  list_endpoints()
  list_endpoints(language="java", type="graphql")
  list_endpoints(language="java", type="rest")
  list_endpoints(language="typescript")
  list_endpoints(language="typescript", type="routes")`,

  args: {
    language: tool.schema
      .enum(["auto", "java", "typescript"])
      .optional()
      .describe("Language to scan. 'auto' runs all registered analyzers. Defaults to 'auto'."),

    type: tool.schema
      .enum(["any", "graphql", "rest", "routes"])
      .optional()
      .describe("Filter by endpoint type. 'graphql' = @QueryMapping/@MutationMapping. 'rest' = HTTP mappings. 'routes' = RR7 loaders/actions/layouts. Defaults to 'any'."),

    includeComponents: tool.schema
      .boolean()
      .optional()
      .describe("Include component-only pages (routes with no loader or action). Default: false. Enable when you need the full route inventory."),
  },

  async execute(args, context) {
    const workspace = context.directory ?? process.cwd()
    const resolvedWorkspace = path.resolve(workspace)

    try {
      await fs.access(SCRIPT)
    } catch {
      return `list_endpoints.py not found at: ${SCRIPT}`
    }

    const language = args.language ?? "auto"
    const type = args.type ?? "any"
    const includeComponents = args.includeComponents ?? false
    const uvDeps = uvDepsForLanguage(language)

    let stdout = ""
    let stderr = ""

    try {
      const uvArgs = uvDeps.length > 0
        ?           ["uv", "run", ...uvDeps, "python3", SCRIPT, resolvedWorkspace, JSON.stringify({ language, type, includeComponents })]
        : ["uv", "run", "python3", SCRIPT, resolvedWorkspace, JSON.stringify({ language, type, includeComponents })]

      const proc = Bun.spawn(uvArgs, {
        stdout: "pipe",
        stderr: "pipe",
        cwd: workspace,
      })

      stdout = await new Response(proc.stdout).text()
      stderr = await new Response(proc.stderr).text()
      const exitCode = await proc.exited

      if (exitCode !== 0) {
        return `list_endpoints failed (exit ${exitCode})\n\n${stderr || stdout}`
      }
    } catch (e) {
      return `Failed to run uv: ${e instanceof Error ? e.message : String(e)}\n\nMake sure 'uv' is installed and in PATH.`
    }

    let result: {
      formatted?: string
      total?: number
      byLanguage?: Array<{ language: string; framework: string; count: number }>
      errors?: unknown[]
      error?: string
      trace?: string
    }

    try {
      result = JSON.parse(stdout.trim())
    } catch {
      return `Could not parse output:\n${stdout}\n\nStderr:\n${stderr}`
    }

    if (result.error) {
      return `Error:\n${result.error}${result.trace ? `\n\nTraceback:\n${result.trace}` : ""}`
    }

    if (!result.formatted || result.total === 0) {
      return `No endpoints found in '${workspace}'.`
    }

    const filterParts: string[] = []
    if (language !== "auto") filterParts.push(`language: ${language}`)
    if (type !== "any") filterParts.push(`type: ${type}`)
    const filterStr = filterParts.length > 0 ? `\nFilters: ${filterParts.join(", ")}` : ""

    const errorNote = result.errors && result.errors.length > 0
      ? `\n\nWarnings: ${result.errors.length} analyzer(s) reported errors.`
      : ""

    return [
      `Workspace: ${path.relative(process.cwd(), resolvedWorkspace) || "."}${filterStr}`,
      "",
      result.formatted,
      errorNote,
    ].join("\n").trim()
  },
})

File: list_endpoints.py
#!/usr/bin/env python3
"""
list_endpoints.py — Thin CLI entrypoint for endpoint/route discovery.

Delegates to endpoints_core for all analysis.
Accepts: <workspace> [options_json]

Options JSON (optional):
  language: string (default "auto") — "auto", "java", "typescript"
  type:     string (default "any")  — "any", "graphql", "rest", "routes"
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

_TOOLS_DIR = str(Path(__file__).resolve().parent)
if _TOOLS_DIR not in sys.path:
    sys.path.insert(0, _TOOLS_DIR)


def _init_analyzers(language: str) -> None:
    """Import only the analyzers needed based on language selection."""
    if language in ("auto", "java"):
        try:
            import tree_sitter_java  # noqa: F401
        except ImportError:
            if language == "java":
                print(
                    json.dumps(
                        {
                            "error": "tree-sitter-java not available. Install with: uv pip install tree-sitter-java"
                        }
                    )
                )
                sys.exit(1)

    if language in ("auto", "typescript"):
        # TypeScript analyzer uses regex only — no tree-sitter needed
        pass

    # Import analyzers — triggers registration into endpoints_core.common.ANALYZERS
    if language == "java":
        from endpoints_core.analyzers import java  # noqa: F401
    elif language == "typescript":
        from endpoints_core.analyzers import typescript  # noqa: F401
    else:
        from endpoints_core.analyzers import java, typescript  # noqa: F401


def main() -> None:
    if len(sys.argv) < 2:
        print(
            json.dumps({"error": "Usage: list_endpoints.py <workspace> [options_json]"})
        )
        sys.exit(1)

    workspace_str = sys.argv[1]
    options: dict = {}
    if len(sys.argv) >= 3:
        try:
            options = json.loads(sys.argv[2])
        except json.JSONDecodeError:
            pass

    language = str(options.get("language", "auto")).lower()
    type_filter = str(options.get("type", "any")).lower()

    _init_analyzers(language)

    from endpoints_core.common import ANALYZERS
    from endpoints_core.formatter import format_results

    workspace = Path(workspace_str).resolve()
    if not workspace.exists():
        print(json.dumps({"error": f"Workspace not found: {workspace_str}"}))
        sys.exit(1)

    results = []
    for lang_id, analyze_fn in ANALYZERS:
        if language != "auto" and lang_id != language:
            continue
        try:
            result = analyze_fn(workspace, options)
            results.append(result)
        except Exception as exc:
            import traceback

            results.append(
                {
                    "language": lang_id,
                    "error": str(exc),
                    "trace": traceback.format_exc(),
                }
            )

    # Separate clean results from errors
    clean = [r for r in results if hasattr(r, "endpoints")]
    errors = [r for r in results if not hasattr(r, "endpoints")]

    total = sum(len(r.endpoints) for r in clean)
    formatted = format_results(clean, type_filter)

    output: dict = {
        "formatted": formatted,
        "total": total,
        "byLanguage": [
            {
                "language": r.language,
                "framework": r.framework,
                "count": len(r.endpoints),
                "endpoints": [
                    {
                        "name": e.name,
                        "kind": e.kind,
                        "file": e.file,
                        "line": e.line,
                        "path": e.path,
                        "group": e.group,
                    }
                    for e in r.endpoints
                ],
            }
            for r in clean
        ],
    }

    if errors:
        output["errors"] = errors

    print(json.dumps(output))


if __name__ == "__main__":
    try:
        main()
    except Exception as exc:
        import traceback

        print(json.dumps({"error": str(exc), "trace": traceback.format_exc()}))
        sys.exit(1)
